### PR TITLE
[Feat]: Implement support for rustls

### LINKED ISSUE #574: feautre request: support rustls
Sometimes I have to cross-compile my project which uses redis with tls feature enabled and it causes a lot of pain and suffering due to openssl issues

It seemd that using rustls instead of native-tls would solve those issues

### PR DESCRIPTION
**Pull request by LeoRowan**
This PR will add support for tls connections using rustls under the "rustls" feature flag. It also adds support for tokio rustls under the "tokio-rustls-comp" flag.

Any feedback is greatly appreciated!
Resolves #574.

### CODE REVIEW

**Comment by LeoRowan** [comment]
@Marwes This is ready for review, thanks!

File: src/aio/tokio.rs [hunk 1]
```diff
@@ -117,6 +127,35 @@ impl RedisRuntime for Tokio {
...
+            .await
+            .map(Tokio::TcpTls)?)
+    }
+    #[cfg(all(feature = "rustls", feature = "tls"))]
+    async fn connect_tcp_tls(
+        _hostname: &str,
+        _socket_addr: SocketAddr,
+        _insecure: bool,
+    ) -> RedisResult<Self> {
+        fail!((
```
Comment (djc) [inline src/aio/tokio.rs:153]: This kind of thing is usually not a good idea. I think we should just pick a preferred implementation and use that one, or make the caller pick one explicitly.

File: src/connection.rs [hunk 2]
```diff
@@ -286,20 +293,49 @@ struct TcpTlsConnection {
...
     #[cfg(feature = "tls")]
     TcpTls(TcpTlsConnection),
+    #[cfg(feature = "rustls")]
+    TcpRustls(TcpRustlsConnection),
     #[cfg(unix)]
     Unix(UnixConnection),
 }
 
+#[cfg(feature = "rustls")]
+struct NoCertificateVerification;
```
Comment (djc) [inline src/connection.rs:322]: I don't think this should be enabled/available by default.

File: Cargo.toml [hunk 3]
```diff
@@ -53,11 +53,16 @@ rand = { version = "0.8", optional = true }
...
 async-trait = "0.1.24"
 
-# Only needed for TLS
+# Only needed for native tls
 native-tls = { version = "0.2", optional = true }
 tokio-native-tls = { version = "0.3", optional = true }
 async-native-tls = { version = "0.4", optional = true }
 
+# Only needed for rustls
+rustls = { version = "0.20.4", features = ["dangerous_configuration"], optional = true }
```
Comment (djc) [inline Cargo.toml:62]: I don't think the `dangerous_configuration` feature should be enabled by default.

**Review by djc — COMMENTED** [review]
I think this mostly looks good, though I don't think we should expose the insecure feature as is. I would also hope we can take the clippy fixes from a different branch and merge them separately to make this easier to review.

**Comment by LeoRowan** [comment]
Sounds good, I mostly added that insecure option by default to match what the `native-tls` feature does. I'll remove the clippy fixes and update the PR with the requested changes.

**Comment by tasn** [comment]
Would love to see it too! 🙏🏻

File: Cargo.toml [hunk 4]
```diff
@@ -53,11 +53,16 @@ rand = { version = "0.8", optional = true }
...
 
-# Only needed for TLS
+# Only needed for native tls
 native-tls = { version = "0.2", optional = true }
 tokio-native-tls = { version = "0.3", optional = true }
 async-native-tls = { version = "0.4", optional = true }
 
+# Only needed for rustls
+rustls = { version = "0.20.4", optional = true }
+webpki-roots = { version = "0.22.3", optional = true }
```
Comment (djc) [inline Cargo.toml:63]: Ideally we would also offer the choice of using rustls-native-certs here instead which I think is preferred in most cases, but feel free to leave that out of this PR.

**Review by djc — COMMENTED** [review]
This seems pretty reasonable. Please rebase and address the comments.

**Comment by LeoRowan** [comment]
> any update on this feature ？

Apologies, I started a new job and haven't had much time for this. I may be able to get back to it, but anyone is welcome to pick up where I left off in the meantime

**Comment by ClSlaid** [comment]
This PR seems pretty stale, may I implement this feature on your work?

**Comment by djc** [comment]
> This PR seems pretty stale, may I implement this feature basing on your work?

The author already gave permission for this:

> Apologies, I started a new job and haven't had much time for this. I may be able to get back to it, but anyone is welcome to pick up where I left off in the meantime

**Comment by ClSlaid** [comment]
> > This PR seems pretty stale, may I implement this feature basing on your work?
> 
> The author already gave permission for this:
> 
> > Apologies, I started a new job and haven't had much time for this. I may be able to get back to it, but anyone is welcome to pick up where I left off in the meantime

Working on it. You may receive a new PR this weekend. :)

**Comment by jaymell** [comment]
Closing in favor of #725

**Pull request closed without merging**